src/ip2name.js

```diff
@@ -0,0 +1,34 @@
+const dns = require('mz/dns');
+
+/**
+ * Get the hostname corresponding to the given IP.  This hostname
+ * must be the only hostname to which that IP resolves in reverse DNS (via
+ * PTR records), and the hostname must only resolve to the given IP.  The
+ * hostname must not contain a wildcard character.  The IP must be an IPv4
+ * address.
+ */
+module.exports = async ip => {
+  // if ip is given in IPv4-translated format, strip the prefix.
+  ip = ip.replace(/^::ffff:/, '');
+
+  let hostnames = await dns.reverse(ip);
+  if (hostnames.length > 1) {
+    throw new Error(`Several hostnames found for ${ip}`);
+  } else if (hostnames.length === 0) {
+    throw new Error(`No hostnames found for ${ip}`);
+  }
+
+  let hostname = hostnames[0];
+  if (!/^[a-zA-Z0-9-.]*$/.test(hostname)) {
+    throw new Error(`Hostname for ${ip} is invalid`);
+  }
+
+  let forward = await dns.resolve4(hostname);
```

Comment:
Right.  (It exists, actually, we just don't use it).

I probably should have included a comment to that effect!  

My reasoning was that v6-only is still a long way away, and a dual-stack host (v4 and v6) would have two IPs and two names, while this code checks that a host has a single name.  We could certainly revisit the question later if we deploy this in a place where we use v6.